extensions/wikia/EmbeddableDiscussions/EmbeddableDiscussionsController.class.php

```diff
@@ -9,161 +9,147 @@ class EmbeddableDiscussionsController {
+	public function __construct() {
 		$this->templateEngine = ( new Wikia\Template\MustacheEngine )->setPrefix( __DIR__ . '/templates' );
 	}
 
-	public static function onParserFirstCallInit( Parser $parser ) {
-		global $wgEnableDiscussions;
+	public function render( array $args ) {
+		global $wgCityId;
 
-		if ( $wgEnableDiscussions ) {
-			$parser->setHook( self::TAG_NAME, [ new self(), 'render' ] );
-		}
+		$parameters = [
+			static::PARAM_MOSTRECENT => $args[ static::PARAM_MOSTRECENT ] ?? false,
+			static::PARAM_SIZE => $args[ static::PARAM_SIZE ] ?? self::ITEMS_DEFAULT,
+			static::PARAM_COLUMNS => $args[ static::PARAM_COLUMNS ] ?? self::COLUMNS_DEFAULT,
+			static::PARAM_CATEGORY => $args[ static::PARAM_CATEGORY ] ?? '',
+		];
 
-		return true;
-	}
+		$errorMessages = $this->validateParameters( $parameters );
+
+		$modelData = ( new DiscussionsThreadModel( $wgCityId ) )
+			->getData( $parameters[ static::PARAM_MOSTRECENT ], $parameters[ static::PARAM_SIZE ], $parameters[ static::PARAM_CATEGORY ] );
 
-	public static function onBeforePageDisplay() {
-		\Wikia::addAssetsToOutput( 'embeddable_discussions_js' );
-		\Wikia::addAssetsToOutput( 'embeddable_discussions_scss' );
+		$categoryError = $this->validateCategory( $args, $modelData );
+		if ( $categoryError ) {
+			$errorMessages[] = $categoryError;
+		}
 
-		JSMessages::enqueuePackage( 'EmbeddableDiscussions', JSMessages::EXTERNAL );
+		if ( !empty( $errorMessages ) ) {
+			return $this->templateEngine
+				->clearData()
+				->setData( [ 'errorMessages' => $errorMessages ] )
+				->render( 'DiscussionError.mustache' );
+		}
 
-		return true;
+		return F::app()->checkSkin( 'wikiamobile' ) ?
+			$this->renderMobile( $modelData, $parameters[ static::PARAM_MOSTRECENT ], $parameters[ static::PARAM_SIZE ] ) :
+			$this->renderDesktop( $modelData, $parameters[ static::PARAM_MOSTRECENT ], $parameters[ static::PARAM_CATEGORY ], $parameters[ static::PARAM_COLUMNS ] );
 	}
 
 	/**
-	 * Checks arguments for errors.
-	 * @param array $args
-	 * @param string errorMessage Return parameter with the proper error message to show. Disregard if return is false
-	 * @return true if ok, false if error
+	 * @param array $parameters
+	 * @return array of error strings or empty array if all args valid
 	 */
-	private function checkArguments( array $args, $modelData, &$errorMessage ) {
-		// mostrecent must be bool
-		if ( isset( $args['mostrecent'] ) &&
-			$args['mostrecent'] !== 'true' &&
-			$args['mostrecent'] !== 'false'
-		) {
-			$errorMessage = wfMessage( 'embeddable-discussions-parameter-error', 'mostrecent' )->plain() .
-				wfMessage( 'embeddable-discussions-parameter-error-boolean' )->plain();
-
-			return false;
+	private function validateParameters( array $parameters ) {
+		$errors = [];
+
+		// PARAM_MOSTRECENT must be bool
+		if ( !AttributesValidator::isBoolish( $parameters[ static::PARAM_MOSTRECENT ] ) ) {
+			$errors[] = wfMessage( 'embeddable-discussions-parameter-error',
```

Comment:
![MINOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png) Define a constant instead of duplicating this literal "embeddable-discussions-parameter-error" 4 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://10.8.74.32:9000/coding_rules#rule_key=php%3AS1192)